astrocut/cube_cut.py

```diff
@@ -815,16 +815,23 @@ def cube_cut(self, cube_file, coordinates, cutout_size,
 
         warnings.filterwarnings("ignore", category=wcs.FITSFixedWarning)
         fits_options: Dict[str, Any] = {"mode": "denywrite", "lazy_load_hdus": True}
+
         if cube_file.startswith("s3://"):
             fits_options["use_fsspec"] = True
-            # block size should be:
-            # block_size <= m * hdul[1].section.shape[2] * hdul[1].section.shape[3] * 4 bytes
             fits_options["fsspec_kwargs"] = {"default_block_size": 10_000, "anon": True}
             # only use .section for remote data
             cube_data_prop = "section"
         else:
             fits_options["memmap"] = True
             cube_data_prop = "data"
+
+        # block size should be:
+        # block_size <= m * hdul[1].section.shape[2] * hdul[1].section.shape[3] * 4 bytes
+        if cube_file.startswith("s3://"):
+            with fits.open(cube_file, **fits_options) as cube:
+                block_size = cube[1].section.shape[1] * cube[1].section.shape[2] * cube[1].section.shape[3] * 4
```

Comment:
i didn't quite think through this entirely and didn't realize we're going to have to read out the header of the cube file each time we make a cutout. i'm wondering how much latency this adds and whether it's worth it versus just taking a hardcoded blocksize approach as we had before. 

can we do some experiments to determine how many requests and how much data is transferred when we set the block size on a per cutout basis (with different cutout sizes and cube sizes) versus just setting it to the same value all the time?  and how much time is added to the cutout by getting the headers here?

Reply:
Yeah I can do that. Originally I thought the intention was to hard-code a default block size for all requests. I thought that was why we were trying to determine the "optimal block size" , so we can hard-code a block size that is optimal in terms of request time while also limiting the # of requests to the server.

Reply:
note, we already do read some table info from the cube in `_parse_table_info`. It would be convenient if we could somehow combine some of this metadata extraction into a single remote read.

Reply:
This happens after the cube file is opened with `**fits_options` tho.

Reply:
yes, i'm aware. i was wondering if we should refactor. i was thinking that maybe we open the file once to read all the metadata that we need.

and then we open it once more to actually make the cutout, with the precise block size.

maybe it's not worth worrying about. i am concerned about all the extra access to these remote files being very slow -- so if there's opportunities to combine those together i think we could benefit.

Reply:
One thing worth mentioning is that making the block size a function of the users input cutout size (as opposed to using a fixed one) adds a few steps extra steps, especially if the cutout size is fed in any unit other than pixels. The script needs to first convert the cutout size to pixels in order to calculate the `block_size` correctly. I added a new method called `_convert_cutout_size` to handle this, and I'll do some analysis tonight on how much time all of these steps add to the request(s). I'll update this thread with the results.

Reply:
OK. FWIW, these are the times I've been seeing with and without computing a custom block size based on input cutout size (when `custom_block_size` is turned off, the block size is just the original 10,000 bytes: 
<img width="1080" alt="image" src="https://user-images.githubusercontent.com/32107699/235518849-95bfa995-dce9-4577-a87d-045c0e9c1c25.png">

Reply:
That indicates to me that the custom block size computation is likely incorrect.

Can you print out the computed block sizes?

How about enabling Geert's context manager that reports on bytes/requests.